### PR TITLE
openems: init at 0.35.0

### PR DESCRIPTION
**Pull request by matthuszagh**
[OpenEMS](http://openems.de/index.php/Main_Page.html) is an open-source electromagnetic field solver that uses the FDTD method.

This commit adds a number of related packages as dependencies for OpenEMS, including appcsxcad, csxcad, fparser, hyp2mat, and qcsxcad.

Additionally, it makes a few changes to existing, dependent packages:

1. Adds QT GUI support to VTK.
2. Adds gl2ps as an optional dependency of octave, which is needed for certain OpenEMS plots.
3. Adds openmpi C++ libraries.
4. Makes cgal headers and libraries discoverable by OpenEMS.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by matthewbauer** [comment]
@grahamcofborg eval

**Comment by matthuszagh** [comment]
Let me know if I need to do something on my end for this. I'm not really sure what that error message means.

**Comment by matthuszagh** [comment]
I've updated this to include the optional python interface.

File: pkgs/applications/science/electronics/openems/default.nix [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    repo = "openEMS";
+    rev = "ffcf5ee0a64b2c64be306a3154405b0f13d5fbba";
+    sha256 = "1c8wlv0caxlhpicji26k93i9797f1alz6g2kc3fi18id0b0bjgha";
+  };
+
+  nativeBuildInputs = [
+    cmake
+  ];
+
+  cmakeFlags = (if withMPI then [ "-DWITH_MPI=ON" ] else []);
```
Comment (jonringer) [inline pkgs/applications/science/electronics/openems/default.nix]:
```suggestion
  cmakeFlags = stdenv.lib.optionals withMPI [ "-DWITH_MPI=ON" ];
```

File: pkgs/development/python-modules/python-csxcad/default.nix [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  version = "0.6.2";
+
+  src = fetchFromGitHub {
+    owner = "thliebig";
+    repo = "CSXCAD";
+    rev = "55899d0fc3dbfc2eb3ec60af9783925926e661a9";
+    sha256 = "18h10575c8k0bx7l0pk5ksqfc1vvnsg98br2987y8x4jmrw2jy50";
+  };
+
+  postUnpack = "sourceRoot+=/python; echo source root reset to $sourceRoot";
```
Comment (jonringer) [inline pkgs/development/python-modules/python-csxcad/default.nix]:
```suggestion
  sourceRoot = "source/python";
```

File: pkgs/development/python-modules/python-openems/default.nix [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  version = "0.0.33";
+
+  src = fetchFromGitHub {
+    owner = "thliebig";
+    repo = "openEMS";
+    rev = "ffcf5ee0a64b2c64be306a3154405b0f13d5fbba";
+    sha256 = "1c8wlv0caxlhpicji26k93i9797f1alz6g2kc3fi18id0b0bjgha";
+  };
+
+  postUnpack = "sourceRoot+=/python; echo source root reset to $sourceRoot";
```
Comment (jonringer) [inline pkgs/development/python-modules/python-openems/default.nix]: sourceRoot = "source/python";

**Review by jonringer — COMMENTED** [review]
to conform to CONTRIBUTING.md, please make it so that roughly 1 commit = 1 package.

```
appcsxcad: init at 0.2.2
csxcad: init at 0.6.2
....
```

**Comment by matthuszagh** [comment]
Changes implemented. This currently fails to build because of qscintilla/octave, so just waiting on [this pr](https://github.com/NixOS/nixpkgs/pull/70391).

**Comment by matthuszagh** [comment]
Octave has been fixed upstream. I've fixed the merge conflict and this now builds correctly.

**Comment by jonringer** [comment]
@GrahamcOfBorg eval

**Comment by matthuszagh** [comment]
Just a heads up, I believe the eval should succeed now that my name is in the maintainers list.

**Comment by matthuszagh** [comment]
@jonringer I've fixed the merge conflicts and got the package working with some other package updates that occurred since the original PR. Please let me know any issues etc. Thanks!

**Comment by jonringer** [comment]
I would break this PR apart into many. As it stands, it would take a lot to review and I think the process would be painful for the both of us.

**Comment by matthuszagh** [comment]
Sure, more than happy to structure this in a way that's easier on you all. How do you suggest I break it up? Roughly, each of those commits should be a separate PR? What's the best way to reference that all these changes are made to support the addition of openems? Just state that and link to the openems webpage? My only concern is that the need for some of these changes may seem less apparent outside the context of openems.

**Comment by matthuszagh** [comment]
@jonringer thanks so much for the guidance on this. I've submitted the [first PR](https://github.com/NixOS/nixpkgs/pull/84644). Others to follow that merge.

**Comment by jonringer** [comment]
@matthuszagh if you want to tag me for reviews, feel free to do so

**Comment by doronbehar** [comment]
There are merge conflicts.

**Comment by matthuszagh** [comment]
@doronbehar yes, this isn't ready to merge yet. I'm resubmitting it as a series of individual PRs. When there is one commit left, I'll rebase.

**Comment by jonringer** [comment]
I think most of these are merged, should we close?

**Comment by matthuszagh** [comment]
Sure, fine by me. Closing.

**Pull request closed without merging**